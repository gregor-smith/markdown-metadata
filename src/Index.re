
type t = Js.Dict.t(array(string));


[@bs.module "markdown-metadata"]
external parseMetadata : string => (t, string) = "parseMetadata";
