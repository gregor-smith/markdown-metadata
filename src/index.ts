function stringIsWhitespace(string: string) {
    if (string.length === 0) {
        return true
    }
    for (let index = 0; index < string.length; index++) {
        if (string.charAt(index) !== ' ') {
            return false
        }
    }
    return true
}


export type Metadata = { [ key: string ]: string[] }


const metadataRegex = /^([a-z_]+): (.+)/i


export function parseMetadata(markdown: string): [ Metadata, string ] {
    const trimmed = markdown.trim()
    const split = trimmed.split('\n')
    const metadata: Metadata = {}

    if (split.length === 0 || !metadataRegex.test(split[0])) {
        return [ metadata, trimmed ]
    }

    let lastKey: string | null = null
    let index = 0
    for (; index < split.length; index++) {
        const line = split[index]
        if (stringIsWhitespace(line)) {
            break
        }
        const match = metadataRegex.exec(line)
        if (match === null) {
            if (lastKey === null) {
                break
            }
            metadata[lastKey].push(line.trim())
        }
        else {
            lastKey = match[1]
            metadata[lastKey] = [ match[2].trim() ]
        }
    }

    const text = split.slice(index + 1).join('\n')
    return [ metadata, text ]
}
