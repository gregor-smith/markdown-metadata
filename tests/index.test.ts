import fs from 'fs'
import path from 'path'

import { parseMetadata } from '../src'


const readExample = (filename: string) => {
    const filePath = path.join(__dirname, 'examples', filename)
    return new Promise<string>((resolve, reject) =>
        fs.readFile(filePath, { encoding: 'utf8' }, (error, data) => {
            if (error !== null) {
                reject(error)
                return
            }
            resolve(data)
        })
    )
}


test('Single line metadata', async () => {
    const example = await readExample('single-line.md')
    const output = parseMetadata(example)
    expect(output).toMatchSnapshot()
})


test('Multi line metadata', async () => {
    const example = await readExample('multi-line.md')
    const output = parseMetadata(example)
    expect(output).toMatchSnapshot()
})


test('Whitespace before metadata', async () => {
    const example = await readExample('whitespace.md')
    const output = parseMetadata(example)
    expect(output).toMatchSnapshot()
})


test('No metadata', async () => {
    const example = await readExample('no-metadata.md')
    const output = parseMetadata(example)
    expect(output).toMatchSnapshot()
})
