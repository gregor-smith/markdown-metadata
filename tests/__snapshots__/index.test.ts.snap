// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multi line metadata 1`] = `
Array [
  Object {
    "date": Array [
      "2019-01-02 12:00",
    ],
    "tags": Array [
      "Programming",
      "Testing",
    ],
    "title": Array [
      "Test title 2",
    ],
  },
  "Dolor sed viverra ipsum nunc aliquet bibendum enim.
In massa tempor nec feugiat.
Nunc aliquet bibendum enim
facilisis gravida.",
]
`;

exports[`No metadata 1`] = `
Array [
  Object {},
  "Egestas egestas fringilla phasellus faucibus
scelerisque eleifend. Sagittis orci a
scelerisque purus semper eget duis.",
]
`;

exports[`Single line metadata 1`] = `
Array [
  Object {
    "date": Array [
      "2019-01-01 12:00",
    ],
    "tag": Array [
      "Programming",
    ],
    "title": Array [
      "Test title",
    ],
  },
  "Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.",
]
`;

exports[`Whitespace before metadata 1`] = `
Array [
  Object {
    "date": Array [
      "2019-01-03 12:00",
    ],
    "tags": Array [
      "Programming",
      "Testing",
      "Whitespace",
    ],
    "title": Array [
      "Test title 3",
    ],
  },
  "Nisl nunc mi ipsum faucibus vitae aliquet nec ullamcorper.
Amet luctus venenatis lectus magna fringilla.
Volutpat maecenas volutpat blandit aliquam etiam erat velit scelerisque in.",
]
`;
